Civi/Schema/SqlEntityStorage.php

```diff
@@ -23,7 +25,65 @@ public function __construct(string $entityName) {
   }
 
   public function writeRecords(array $records): array {
-    // TODO: Implement writeRecords() method.
+    $results = [];
+    foreach ($records as $record) {
+      $results[] = $this->writeRecord($record);
+    }
+    return $results;
+  }
+
+  /**
+   * Create or update a record from supplied params.
+   *
+   * If 'id' is supplied, an existing record will be updated
+   * Otherwise a new record will be created.
+   *
+   * @param array $record
+   *
+   * @return static
+   * @throws \CRM_Core_Exception
+   */
+  public function writeRecord(array $record): object {
+    // Todo: Support composite primary keys
+    $idField = CoreUtil::getIdFieldName($this->entityName);;
+    $op = empty($record[$idField]) ? 'create' : 'edit';
+
+    // For legacy reasons, empty values would sometimes be passed around as the string 'null'.
+    // The DAO treats 'null' the same as '', and an empty string makes a lot more sense!
+    // For the sake of hooks, normalize these values.
+    $record = array_map(function ($value) {
+      return $value === 'null' ? '' : $value;
+    }, $record);
+
+    \CRM_Utils_Hook::pre($op, $this->entityName, $record[$idField] ?? NULL, $record);
+    $entityDAO = EntityRepository::getClass($this->entityName);
+    $fields = $entityDAO::getSupportedFields();
+    $instance = new $entityDAO();
+    // Ensure fields exist before attempting to write to them
+    $values = array_intersect_key($record, $fields);
+    $instance->copyValues($values);
+    if (empty($values[$idField]) && array_key_exists('name', $fields) && empty($values['name'])) {
+      $instance->makeNameFromLabel();
+    }
+    if (empty($values[$idField]) && array_key_exists('frontend_title', $fields) && empty($values['frontend_title'])) {
+      $instance->frontend_title = $instance->title;
+    }
+    if (empty($values[$idField]) && array_key_exists('frontend_title', $fields) && !$instance->frontend_title) {
+      // Still empty? Fall back to name.
+      $instance->frontend_title = $instance->name;
+    }
+    if (empty($values[$idField]) && array_key_exists('title', $fields) && empty($values['title']) && array_key_exists('frontend_title', $fields) && $instance->frontend_title) {
+      $instance->title = $instance->frontend_title;
+    }
```

Comment:
Oof, this frontend_title stuff feels hacky.
I think let's try to keep this function pure, and if specific entities need this hack they can implement `hook_civicrm_pre()`